packages/@aws-cdk/toolkit-lib/lib/api/cloudformation/template-body-parameter.ts

```diff
@@ -99,9 +99,21 @@ export async function makeBodyParameter(
  *
  * Replaces environment placeholders (which this field may contain),
  * and reformats s3://.../... urls into S3 REST URLs (which CloudFormation
- * expects)
+ * expects).
+ *
+ * We need to return the official region- and partition-specific URL for AWS S3
+ * here, so we use the SDK's information about endpoints. At the same time, the
+ * SDK allows overriding this URL by setting an environment variable
+ * (specifically $AWS_ENDPOINT_URL_S3) but we want to *not* honor that, because
+ * there's a 99.9% chance this URL will not be routable from AWS CloudFormation.
+ *
+ * To allow for the off chance that someone is running this tool against a
+ * custom build of CloudFormation that does need a specific S3 endpoint passed
+ * to it, we'll introduce a new environment variable that we'll respect instead:
+ *
+ *  AWS_ENDPOINT_URL_S3_FOR_CLOUDFORMATION
```

Comment:
(blocking): I'm leaning against the proliferation of introducing undocumented env vars doing things. I'm guessing this was added to preempt new issues?  Let me reason this through:
- Currently when `AWS_ENDPOINT_URL_S3` is set it will break deployment
- This PR fixes that
- If someone deliberately used `AWS_ENDPOINT_URL_S3` to do _something_ unknown that we do not know about, they will be broken now.
- How do they find the new env var? I say they won't and they will open an issue anyway.

I vote we only address this if someone has this problem.